### ISSUE TITLE
Solution for updating (re-rendering) the button-card

### ISSUE THREAD

**Issue by il77781** [thomasloven/hass-browser_mod#757]
Hello!
To be honest, I'm not sure which type of issue I should have chosen...
In general, the problem is that if you use `browser_mod.popup` to open a [button-card](https://github.com/custom-cards/button-card), which uses, for example, templates that change the state (for example, styles) of buttons when (again, for example) you press these buttons, nothing actually happens to the popup - when you press the buttons, its state in popup does not change...
Yes, this topic has already been discussed, at least here - https://github.com/custom-cards/button-card/issues/397 and here - https://github.com/thomasloven/hass-browser_mod/issues/186... 
But, unfortunately, no solution, as far as I was able to find out through the search, was never found - the developers nodded at each other - and, unfortunately, everything stopped there...
It should be noted that the problem with the lack of "re-rendering" is not observed for all types of cards...
As far as I understand (please correct me if necessary), the problem is that when the popup is open and when clicking on the button-card buttons, which trigger certain events, in turn, affecting the values calculated by the templates, the popup itself has already been formed... And changes in the values calculated by the templates simply do not "reach" the popup...
If this is the case, then maybe it is possible to add an additional service with the conditional name `browser_mod.popup_re-render`, which will "redraw" the current popup?
If the statement from this comment - https://github.com/custom-cards/button-card/issues/397#issuecomment-724126534 is more likely and easier to implement, then maybe something can be done about it?
In the event that a solution has already been found, and I just haven't found it, please direct me in the right direction...

**Comment by dcapslock** [comment]
Hi @il77781 do you have a clear example of config that shows the issue?

**Comment by il77781** [comment]
@dcapslock, hello!
Unfortunately, I have already deleted this config as non-working...
But I will try to recreate it soon and post it here...

**Comment by il77781** [comment]
@dcapslock, hello!
Here is an example of a config that demonstrates the issue...
```
type: custom:button-card
triggers_update: all
styles:
  card:
    - padding: 10px 10px
  grid:
    - grid-template-areas: " \"temperature wind_speed\" \"humidity wind_direction\" "
    - grid-template-columns: 1fr 1fr
    - grid-template-rows: auto auto
    - gap: 10px 2px
  custom_fields:
    temperature:
      - justify-self: start
      - font-size: 14px
    wind_speed:
      - justify-self: start
      - font-size: 14px
    humidity:
      - justify-self: start
      - font-size: 14px
custom_fields:
  temperature: |
    [[[
      return `<ha-state-icon icon="mdi:thermometer" style="height: 20px; color: var(--state-icon-color); background-color: color-mix(in srgb, var(--state-icon-color) 20%, transparent); border-radius: 13px; padding: 3px;"></ha-state-icon>
      <span>Temperature: ${states['sensor.gw1200c_outdoor_temperature'].state} °C</span>`
    ]]]
  wind_speed: |
    [[[
      return `<ha-state-icon icon="mdi:weather-windy" style="height: 20px; color: var(--state-icon-color); background-color: color-mix(in srgb, var(--state-icon-color) 20%, transparent); border-radius: 13px; padding: 3px; rotate: ${states['sensor.gw1200c_wind_direction'].state - 270}deg"></ha-state-icon>
      <span>Wind speed: ${states['sensor.gw1200c_wind_speed'].state} m/s</span>`
    ]]]
  humidity: |
    [[[
      return `<ha-state-icon icon="mdi:water-percent" style="height: 20px; color: var(--state-icon-color); background-color: color-mix(in srgb, var(--state-icon-color) 20%, transparent); border-radius: 13px; padding: 3px;"></ha-state-icon>
      <span>Humidity: ${states['sensor.gw1200c_humidity'].state} %</span>`
    ]]]
  wind_direction:
    card:
      type: custom:button-card
      triggers_update: all
      entity: sensor.gw1200c_wind_direction
      layout: icon_name_state
      name: Wind direction
      show_state: true
      styles:
        card:
          - padding: 0px
          - border: none
        icon:
          - height: 20px
          - width: 20px
          - color: var(--state-icon-color)
          - background-color: color-mix(in srgb, var(--state-icon-color) 20%, transparent)
          - border-radius: 50%
          - padding: 3px
          - left: "-35%"
          - rotate: |
              [[[
                return (states['sensor.gw1200c_wind_direction'].state - 90) + 'deg'
              ]]]
        name:
          - font-size: 14px
          - position: absolute
          - left: 15%
          - overflow: visible
      tap_action:
        action: fire-dom-event
        browser_mod:
          service: browser_mod.popup
          data:
            content:
              type: custom:button-card
              triggers_update: all
              styles:
                card:
                  - padding: 10px 10px
                grid:
                  - grid-template-areas: " \"24h week month\" \"wind_rose wind_rose wind_rose\" "
                  - grid-template-columns: 1fr 1fr 1fr
                  - gap: 5px
              custom_fields:
                24h:
                  card:
                    type: custom:button-card
                    triggers_update: all
                    name: 24 Hours
                    styles:
                      card:
                        - box-shadow: 1px 1px 1px
                    tap_action:
                      action: call-service
                      service: input_number.set_value
                      data:
                        entity_id: input_number.period
                        value: 24
                    state:
                      - operator: template
                        value: |
                          [[[
                            return states['input_number.period'].state === '24.0'
                          ]]]
                        styles:
                          card:
                            - background-color: rgb(0, 188, 212)
                            - box-shadow: inset 2px 2px 2px rgb(0, 80, 90)
                week:
                  card:
                    type: custom:button-card
                    triggers_update: all
                    name: Week
                    styles:
                      card:
                        - box-shadow: 1px 1px 1px
                    tap_action:
                      action: call-service
                      service: input_number.set_value
                      data:
                        entity_id: input_number.period
                        value: 168
                    state:
                      - operator: template
                        value: |
                          [[[
                            return states['input_number.period'].state === '168.0'
                          ]]]
                        styles:
                          card:
                            - background-color: rgb(0, 188, 212)
                            - box-shadow: inset 2px 2px 2px rgb(0, 80, 90)
                month:
                  card:
                    type: custom:button-card
                    triggers_update: all
                    name: Month
                    styles:
                      card:
                        - box-shadow: 1px 1px 1px
                    tap_action:
                      action: call-service
                      service: input_number.set_value
                      data:
                        entity_id: input_number.period
                        value: 720
                    state:
                      - operator: template
                        value: |
                          [[[
                            return states['input_number.period'].state === '720.0'
                          ]]]
                        styles:
                          card:
                            - background-color: rgb(0, 188, 212)
                            - box-shadow: inset 2px 2px 2px rgb(0, 80, 90)
                wind_rose:
                  card:
                    type: custom:windrose-card
                    title: |
                      [[[
                        return 'Wind Rose - ' + states['input_number.period'].state.slice(0,-2) + ' hours'
                      ]]]
                    wind_direction_entity:
                      entity: sensor.gw1200c_wind_direction
                      use_statistics: false
                    windspeed_entities:
                      - entity: sensor.gw1200c_wind_speed
                        name: Wind Speed
                        use_statistics: false
                        render_relative_scale: false
                        speed_unit: auto
                        current_speed_arrow: true
                      - entity: sensor.gw1200c_wind_gust
                        name: Wind Gust
                        use_statistics: false
                        render_relative_scale: false
                        speed_unit: auto
                        current_speed_arrow: true
                    refresh_interval: 300
                    data_period:
                      hours_to_show: |
                        [[[
                          return states['input_number.period'].state
                        ]]]
                    windspeed_bar_location: bottom
                    center_calm_percentage: false
                    direction_labels:
                      cardinal_direction_letters: NESW
                      show_cardinal_directions: true
                      show_intercardinal_directions: true
                      show_secondary_intercardinal_directions: true
                      cardinal_directions_text_size: 60
                      intercardinal_directions_text_size: 45
                      secondary_intercardinal_directions_text_size: 30
                    current_direction:
                      show_arrow: true
                      hide_direction_below_speed: 0
                    actions:
                      windrose:
                        tap_action:
                          entity: sensor.gw1200c_wind_direction
                          action: more-info
                    matching_strategy: full-time
            style: |
              --popup-min-width: 400px
```
If necessary, I can try to shorten the code further...
The card placed on the dashboard looks like this:

![Image](https://github.com/user-attachments/assets/a2d7b679-2df2-4b7a-83c3-5d0024b01522)

When you click on the `wind direction`, a card like this pops up:

![Image](https://github.com/user-attachments/assets/0e840fb3-fea3-476a-95db-2e99b6ac5e90)

As you can see, there are three buttons on it that control the period for which the data is displayed on the wind rose.
Due to the "locality" of the use of variables in the `button-card`, I managed to implement the "switching" of this period only using the "external" `input_number` (if this can be implemented in any other way, I would appreciate a hint)
But, in general, a card with 3 buttons pops up, **but nothing happens when you click on these buttons**: the "pressed" buttons do not become active (the background color of these buttons does not change) and the period for displaying data on the wind rose (as well as the corresponding graph displaying data for this period) does not change! At the same time, the actions assigned to these buttons occur - the `input_number` value **changes immediately** when the buttons are pressed! **Nothing changes on the pop-up card only!** And if I then close the pop-up card and open it again, then everything will be displayed correctly, in accordance with the current `input_number` value...
If I place on the dashboard only the card corresponding to the pop-up card, 
```
type: custom:button-card
styles:
  card:
    - padding: 10px 10px
  grid:
    - grid-template-areas: " \"24h week month\" \"wind_rose wind_rose wind_rose\" "
    - grid-template-columns: 1fr 1fr 1fr
    - gap: 5px
custom_fields:
  24h:
    card:
      type: custom:button-card
      name: 24 Hours
      styles:
        card:
          - box-shadow: 1px 1px 1px
      tap_action:
        action: call-service
        service: input_number.set_value
        data:
          entity_id: input_number.period
          value: 24
      state:
        - operator: template
          value: |
            [[[
              return states['input_number.period'].state === '24.0'
            ]]]
          styles:
            card:
              - background-color: rgb(0, 188, 212)
              - box-shadow: inset 2px 2px 2px rgb(0, 80, 90)
  week:
    card:
      type: custom:button-card
      name: Week
      styles:
        card:
          - box-shadow: 1px 1px 1px
      tap_action:
        action: call-service
        service: input_number.set_value
        data:
          entity_id: input_number.period
          value: 168
      state:
        - operator: template
          value: |
            [[[
              return states['input_number.period'].state === '168.0'
            ]]]
          styles:
            card:
              - background-color: rgb(0, 188, 212)
              - box-shadow: inset 2px 2px 2px rgb(0, 80, 90)
  month:
    card:
      type: custom:button-card
      name: Month
      styles:
        card:
          - box-shadow: 1px 1px 1px
      tap_action:
        action: call-service
        service: input_number.set_value
        data:
          entity_id: input_number.period
          value: 720
      state:
        - operator: template
          value: |
            [[[
              return states['input_number.period'].state === '720.0'
            ]]]
          styles:
            card:
              - background-color: rgb(0, 188, 212)
              - box-shadow: inset 2px 2px 2px rgb(0, 80, 90)
  wind_rose:
    card:
      type: custom:windrose-card
      title: |
        [[[
          return 'Wind Rose - ' + states['input_number.period'].state.slice(0,-2) + ' hours'
        ]]]
      wind_direction_entity:
        entity: sensor.gw1200c_wind_direction
        use_statistics: false
      windspeed_entities:
        - entity: sensor.gw1200c_wind_speed
          name: Wind Speed
          use_statistics: false
          render_relative_scale: false
          speed_unit: auto
          current_speed_arrow: true
        - entity: sensor.gw1200c_wind_gust
          name: Wind Gust
          use_statistics: false
          render_relative_scale: false
          speed_unit: auto
          current_speed_arrow: true
      refresh_interval: 300
      data_period:
        hours_to_show: |
          [[[
            return states['input_number.period'].state
          ]]]
      windspeed_bar_location: bottom
      center_calm_percentage: false
      direction_labels:
        cardinal_direction_letters: NESW
        show_cardinal_directions: true
        show_intercardinal_directions: true
        show_secondary_intercardinal_directions: true
        cardinal_directions_text_size: 60
        intercardinal_directions_text_size: 45
        secondary_intercardinal_directions_text_size: 30
      current_direction:
        show_arrow: true
        hide_direction_below_speed: 0
      actions:
        windrose:
          tap_action:
            entity: sensor.gw1200c_wind_direction
            action: more-info
      matching_strategy: full-time
```
**then everything works correctly:** the "pressed" buttons become active (the background color of these buttons changes) and the period for the data display on the wind rose (as well as the corresponding graph showing the data for this period) is changing correctly!
**In other words, changes do not occur only on the pop-up card - its contents are not updated**. 
Perhaps this has something to do with the "features" of updating the `button-card` (although, as you can see, I have placed `triggers_update: all` wherever possible - but the presence of these lines still does not affect anything), but still... 
Maybe it is possible to add another additional action (service) to update the contents of already "surfaced" (currently open) pop-up cards?

**Comment by dcapslock** [comment]
@il77781 thanks. This has helped me replicate your issue. The core of the issue is that when using a custom: button-card event to popup a popup which has custom: button-cards, then the issue happens. However if a standard Home Assistant card is used, then all works well. 

A workaround here, and perhaps better to manage, is to place your popup in a 'Popup Card' tied to wind direction.

So 

1. wrap your card on dashboard with custom: popup-card.
```
type: custom:popup-card
entity: sensor.gw1200c_wind_direction
card:
  type: custom:button-card
  styles:
    card:
      - padding: 10px 10px
    grid:
      - grid-template-areas: " \"24h week month\" \"wind_rose wind_rose wind_rose\" "
      - grid-template-columns: 1fr 1fr 1fr
      - gap: 5px
  custom_fields:
    24h:
...
```

2. Your wind_direction tap action on your overview card simply becomes
```
      tap_action:
        action: more-info
        entity: sensor.gw1200c_wind_direction
```

Workaround aside, I will look to see whether the issue can be easily tracked down.

**Comment by il77781** [comment]
@dcapslock, yes, it works like this, thank you very much!
Do I understand correctly that this "redefinition" (replacement) of the `more-info` dialog occurs only for the dashboard tab (view) on which the `custom:popup-card` itself is located? Will the standard `more-info` dialog open on other tabs (views) of even the same dashboard?
And, of course, this phrase is a little confusing:
>_Note: It's advisable to use a fire-dom-event tap action instead as far as possible. Popup card is for the few cases where that's not possible. See [services](https://github.com/thomasloven/hass-browser_mod/blob/master/documentation/services.md) for more info._

What is the reason for this?
> Workaround aside, I will look to see whether the issue can be easily tracked down.

I am very grateful to you for your attempts to find a solution!

**Comment by dcapslock** [comment]
> [@dcapslock](https://github.com/dcapslock), yes, it works like this, thank you very much! Do I understand correctly that this "redefinition" (replacement) of the `more-info` dialog occurs only for the dashboard tab (view) on which the `custom:popup-card` itself is located? Will the standard `more-info` dialog open on other tabs (views) of even the same dashboard? And, of course, this phrase from here https://github.com/thomasloven/hass-browser_mod is a little confusing:
> 

After 2.3.4, if there is a Popup Card for an entity within the dashboard then it will be used on another view. If you need to fine tune this approach, you can always create a helper entity that does nothing and use that as the link. 

> > _Note: It's advisable to use a fire-dom-event tap action instead as far as possible. Popup card is for the few cases where that's not possible. See [services](https://github.com/thomasloven/hass-browser_mod/blob/master/documentation/services.md) for more info._
> 
> What is the reason for this?

Not sure. Likely that `more-info` approach is interception HA messge fired in the DOM. I can't see any reason that Popup Card is not a good way to work.

> 
> > Workaround aside, I will look to see whether the issue can be easily tracked down.
> 
> I am very grateful to you for your attempts to find a solution!

I am to the point that it is an issue with how button-card calls the browser mod popup, in that the popups javascript templates are missing. Still to work out why?

**Comment by il77781** [comment]
@dcapslock,

> After 2.3.4, if there is a Popup Card for an entity within the dashboard then it will be used on another view.

I currently have version 2.3.3 installed (and for some reason there are no notifications about the availability of a newer version) - and at the moment, my replacement affects only the dashboard tab on which the `custom:popup-card` itself is located... And it seems to me that this approach is the most logical and usable..

> If you need to fine tune this approach, you can always create a helper entity that does nothing and use that as the link.

Yes, I have already thought of such a solution to the problem (if necessary)...

> Not sure. Likely that `more-info` approach is interception HA messge fired in the DOM. I can't see any reason that Popup Card is not a good way to work.

I understand you... So, I will use it if no other solution is found in the future...

> Still to work out why?

Yes, that would be great!

**Comment by dcapslock** [comment]
> [@dcapslock](https://github.com/dcapslock),
> 
> > After 2.3.4, if there is a Popup Card for an entity within the dashboard then it will be used on another view.
> 
> I currently have version 2.3.3 installed (and for some reason there are no notifications about the availability of a newer version) - and at the moment, my replacement affects only the dashboard tab on which the `custom:popup-card` itself is located... And it seems to me that this approach is the most logical and usable..
> 

Version 2.3.4 is on my 'Phoenix' fork, which was made as lots of PRs were waiting for merge and there has been no response from the maintainer. See the 2.3.4 [README](https://github.com/dcapslock/hass-browser_mod/blob/v2.3.4/README.md#phoenix-version). This version had a fix to be more reliable in handling edit mode of Popup Card. A side effect was the availablbility of a Popup Card to be found in the current dashboard view. I saw this as a feature so left, but looks like maybe not wanted in your case.

**Comment by dcapslock** [comment]
> > I am very grateful to you for your attempts to find a solution!
> 
> I am to the point that it is an issue with how button-card calls the browser mod popup, in that the popups javascript templates are missing. Still to work out why?

The issue is with button-card in that the template config sent with the tap event is empty.

**Comment by il77781** [comment]
> Version 2.3.4 is on my 'Phoenix' fork, which was made as lots of PRs were waiting for merge and there has been no response from the maintainer.

Ah, that's the thing! And I didn't even know about it...

> The issue is with button-card in that the template config sent with the tap event is empty.

And what can be done about it? Do I need to address this issue to the creator of the button-card?

**Comment by dcapslock** [comment]
> And what can be done about it? Do I need to address this issue to the creator of the button-card?

Yes. Though the repo looks quite stale so I don't expect any action. It might be best to stick with a workaround with browser_mod. If 2.3.4 way of finding popup cards across the whole dashboard is limiting, we may be able to look at that.

**Comment by il77781** [comment]
> Though the repo looks quite stale

Do you mean the `button-card` repository?

**Comment by dcapslock** [comment]
> > Though the repo looks quite stale
> 
> Do you mean the `button-card` repository?

Yes. More on the issue as well. What is actually happening is that button_card is updating the templates on itself, so what it passes to browser mod popup service is the template values at the time of calling, not the actual template. So it would be plausible, and I will test, that the actual template you need to include in browser_mod popup service config is a string that then will get passed for evluation in the popup.

**Comment by dcapslock** [comment]
> 
> Yes. More on the issue as well. What is actually happening is that button_card is updating the templates on itself, so what it passes to browser mod popup service is the template values at the time of calling, not the actual template. So it would be plausible, and I will test, that the actual template you need to include in browser_mod popup service config is a string that then will get passed for evluation in the popup.

I can't see any workaround. button_card would need to be adjusted to not evluate templates of config of service calls. 

For your best workaround, it may be to look at how you are displaying the summary card, and use another method including latest in-the-box Home Assistant dashboard features, or say mushroom cards. Your popup can still be button_card, but any config of nested button_card with template will have this issue.

**Comment by il77781** [comment]
> Yes. More on the issue as well. What is actually happening is that button_card is updating the templates on itself, so what it passes to browser mod popup service is the template values at the time of calling, not the actual template. So it would be plausible, and I will test, that the actual template you need to include in browser_mod popup service config is a string that then will get passed for evluation in the popup.

About seven months ago, the author of the button-card replied (on another question), so I think anything is possible...
In this regard, of course, it would be very desirable if you could describe what is happening most accurately... And what needs to be corrected...
I just want to say that the author has already provided such an answer to a similar (in my opinion) problem:
https://github.com/custom-cards/button-card/issues/397#issuecomment-724126534
Maybe if you provide him with more detailed information, the answer and reaction may be different

**Comment by dcapslock** [comment]
> About seven months ago, the author of the button-card replied (on another question), so I think anything is possible... In this regard, of course, it would be very desirable if you could describe what is happening most accurately... And what needs to be corrected... I just want to say that the author has already provided such an answer to a similar (in my opinion) problem: [custom-cards/button-card#397 (comment)](https://github.com/custom-cards/button-card/issues/397#issuecomment-724126534) Maybe if you provide him with more detailed information, the answer and reaction may be different

Simply, any template on the entrie card config is updated on entity changes, including for non-active config for service calls, like browser_mod. This means that the config passed is the template evaluated value, not the template config, which is needed in your case as you have a nested button_card. An option is needed such that templates are not evaluated but passed as config to service calls.

**Comment by il77781** [comment]
> For your best workaround, it may be to look at how you are displaying the summary card, and use another method including latest in-the-box Home Assistant dashboard features, or say mushroom cards. Your popup can still be button_card, but any config of nested button_card with template will have this issue.

Specifically, this problem was solved in a different way - by adding buttons that determine the period of data display, by the author of `custom:windrose-card` ...
But I do not exclude that it may occur in other cases... Therefore, of course, it would be desirable to fix it ...
And as a last resort, of course, I can use the method that you suggested...

**Comment by il77781** [comment]
> Simply, any template on the entrie card config is updated on entity changes, including for non-active config for service calls, like browser_mod. This means that the config passed is the template evaluated value, not the template config, which is needed in your case as you have a nested button_card. An option is needed such that templates are not evaluated but passed as config to service calls.

And do I understand correctly that adding a new service (action) that will update (redraw) an already open popup card will not help in this case?

**Comment by dcapslock** [comment]
> And do I understand correctly that adding a new service (action) that will update (redraw) an already open popup card will not help in this case?

Correct. The template values password to browser_mod simply have `true` or `false` as evaluated at time of button_card calling the browser_mod popup service. So it is not a issue of refresh, its an issue of the config passed.

**Comment by il77781** [comment]
@dcapslock, it seems to be becoming more and more clear to me what the problem is...
I will still try to open the issue in the `button-card` repository and ask to correct the behavior of this card...
If the author provides some kind of answer and it will be related to the `browser_mod.popup`, I will duplicate it here ...
Thank you so much for your attention to my problem and your time!

**Comment by dcapslock** [comment]
Clsoing. If `button-card` investigation is pointing to Browser Mod then please update here. Also `state-switch` may also help your situation. Upcoming 2.4.0 release (in pre-release) will make `state-switch` work again with Broswer Mod.

**Comment by dcapslock** [comment]
@il77781 while looking at another issue, I thought of an alternative solution to templates with custom:button-card.

In the end it is as simple as having the template return a template, to then be used when the browser_mod.popup call is made. Example below.

```yaml
type: custom:button-card
entity: input_select.test_select
name: |
  [[[
     return entity.state
  ]]]
style: null
tap_action:
  action: perform-action
  perform_action: browser_mod.popup
  browser_id: THIS
  data:
    content:
      type: custom:button-card
      entity: input_select.test_select
      name: |
        [[[
           let template = "[[[ return entity.state ]]]";
           return template;
        ]]]
```

**Comment by il77781** [comment]
@dcapslock, hello.
Sorry I haven't replied for so long - I've been busy with some sorrow affairs and I haven't had time to try to deal with this problem.
I'll try to do it as soon as possible...

**Comment by dcapslock** [comment]
No worries. The solution is there when you are ready. Confirmed working fine in the other issue I mentioned this issue in. Z
